---
title: "Cancer genome sequencing report for precision oncology - `r sample_name`"
author: "NCGC Bioinformatics - Lars B Aaasheim, Ghislain Fournous, Daniel Vodak, Sigve Nakken & Eivind Hovig"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: bibliography.bib
output: 
  html_document:
    toc: true
    theme: default
    number_sections: false
    toc_float: true
    toc_depth: 4
params:
  tier1_report: FALSE
  tier2_report: FALSE
  tier3_report: FALSE
  tier4_report: FALSE
  signature_report: FALSE
  cnv_report: FALSE
  snapshot: !r as.POSIXct("2015-01-01 12:30:00")
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
missing_signature_data <- FALSE
```

## Introduction
The tumor case was sequenced using a whole-exome approach with a matching control. Somatic SNVs/InDels were called using both Strelka and MuTect. SNVs that have been called by both MuTect and Strelka are annotated with __STRONG__ _CALL\_CONFIDENCE_, SNV calls made by individual algorithms only are annotated as __WEAK__. Mutational signature analysis has been done with all calls, both weak and strong.

## Somatic SNVs/InDels

### Tier 1 - genomic biomarkers for diagnosis, prognosis, and drug sensitivity {.tabset}

- A total of __`r nrow(variants_tier1)`__ unique, somatic variant(s) in the tumor sample can be mapped to mutation-specific genomic biomarkers in the [database for clinical interpretations of variants in cancer, CIViC](https://civic.genome.wustl.edu/#/home), with a the following number of evidence items:
    * Tier 1A: __`r nrow(clinical_evidence_items_tier1A)`__ evidence items linked to specific amino acid changes
    * Tier 1B: __`r nrow(clinical_evidence_items_tier1B)`__ evidence items linked to codon alterations
    * Tier 1C: __`r nrow(clinical_evidence_items_tier1C)`__ evidence items linked to exon alterations

<br>
```{r conditional_print1, child='tier1.Rmd', eval = tier1_report}
```

### Tier 2 - other cancer mutation hotspots or predicted driver mutations

- A total of __`r nrow(variants_tier2)`__ unique, somatic variant(s) in the tumor sample are predicted as driver mutations or occur in previously detected mutation hotspots
<br>
```{r conditional_print2, child='tier2.Rmd', eval = tier2_report}
```

### Tier 3 - other coding mutations in oncogenes/tumor suppressors/cancer census genes

- A total of __`r nrow(variants_tier3)`__ unique, somatic variant(s) in the tumor sample found within known oncogenes, tumor suppressor genes or cancer census genes:
<br>
```{r conditional_print3, child='tier3.Rmd', eval = tier3_report}
```

<br>

### Tier 4 - other coding mutations

- A total of __`r nrow(variants_tier4)`__ unique, somatic variant(s) are also found in the tumor sample:
<br>
```{r conditional_print4, child='tier4.Rmd', eval = tier4_report}
```
<br>

## Somatic CNA analysis

```{r conditional_print5, child='cnv_report.Rmd', eval = cnv_report}
```
<br>

## Mutational signatures

```{r conditional_print6, child='mutational_signature.Rmd', eval = signature_report}
```

```{r conditional_print7, eval = missing_signature_data}

```

## References
