The annotated variants are based on the following underlying tools/knowledge sources:

* [VEP v85](http://www.ensembl.org/info/docs/tools/vep/index.html) - Variant Effect Predictor release 85 (GENCODE v19 as the gene model)
* [COSMIC v78](http://cancer.sanger.ac.uk/cosmic/) - catalogue of somatic mutations in cancer (Sep 2016)
* [dBNSFP v3.2](https://sites.google.com/site/jpopgen/dbNSFP) - database of non-synonymous functional predictions (March 2016)
* [ICGC v22](https://dcc.icgc.org/) - Somatic mutations discovered in all ICGC (International Cancer Genomics Consortium) tumor cohorts (Aug 2016)
* [ExAC r0.3.1](http://exac.broadinstitute.org/) - germline variant frequencies exome-wide (March 2016)
* [dbSNP b147](http://www.ncbi.nlm.nih.gov/SNP/) - database of short genetic variants (April 2016)
* [1000Genomes phase3](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/) - germline variant frequencies genome-wide (May 2013)
* [ClinVar](http://www.ncbi.nlm.nih.gov/clinvar/) - database of clinically related variants (Nov 2016)
* [DoCM](http://docm.genome.wustl.edu) - database of curated mutations (v3.2, April 2016)
* [CIViC](http://civic.genome.wustl.edu) - clinical interpretations of variants in cancer (Nov 12th 2016)
* [CBMDB](http://www.cancergenomeinterpreter.org/biomarkers) - Cancer Biomarkers database (Oct 28th 2016)
* [IntOGen catalog of driver mutations](https://www.intogen.org/downloads) - (May 2016)
* [Cancer Hotspots](http://cancerhotspots.org) - a resource for statistically significant mutations in cancer (2016)
* [UniProt/SwissProt KnowledgeBase 2016_09](http://www.uniprot.org) - resource on protein sequence and functional information (Sep 2016)
* [Pfam v30](http://pfam.xfam.org) - database of protein families and domains (June 2016)
* [DGIdb](http://dgidb.genome.wustl.edu) - database of interactions betweeen antineoplastic drugs and human proteins (v2.22, February 2016)
* [TSGene v2.0](http://bioinfo.mc.vanderbilt.edu/TSGene/) - tumor suppressor/oncogene database (November 2015)
