---
title: "Cell line sequencing report - `r sample_id`"
author: "NCGC Bioinformatics - Lars B Aaasheim, Ghislain Fournous, Daniel Vodak, Sigve Nakken & Eivind Hovig"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: bibliography.bib
output: 
  html_document:
    toc: true
    theme: default
    number_sections: false
    toc_float: true
    toc_depth: 4
    fig_width: 5
    fig_height: 4
params:
  tier1_report: FALSE
  tier2_report: FALSE
  tier3_report: FALSE
  tier4_report: FALSE
  show_coverage_stats: FALSE
  show_data_sources: TRUE

  snapshot: !r as.POSIXct("2015-01-01 12:30:00")
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = F,warning=FALSE, dpi=300)
```

## Introduction
The tumor cell line was sequenced and analyzed against an artificial control sample (normal). Somatic SNVs/InDels were called using both [Strelka](https://sites.google.com/site/strelkasomaticvariantcaller/home) [@Saunders2012-si] and [MuTect](http://archive.broadinstitute.org/cancer/cga/mutect) [@Cibulskis2013-nr]. SNVs that have been called by both MuTect and Strelka are annotated with __STRONG__ _CALL\_CONFIDENCE_, SNVs called by individual algorithms only are annotated as __WEAK__. InDels were called with Strelka only, and they are thus annotated with __STRONG__ confidence.

### Filtering of calls
The complete set of detected variants by MuTect and Strelka constitue a mix of both somatic and germline calls (due to the lack of a matching control sample). In this report, a filtering/enrichment procedure of the result set has been set up to i) exclude the majority of germline variants, and ii) enrich for cancer-associated variation:

* Exclusion of calls that coincide with germline variants found in [Exome Aggregation Consortium]((http://exac.broadinstitute.org/)) (minor allele frequency >= 0.1% in any population)
* Exclusion of calls that coincide with germline variants found in [1000 Genomes Project](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/) (minor allele frequency >= 1% in any population)
* Exclusion of calls that coincide with variants found in [dbSNP build 147](http://www.ncbi.nlm.nih.gov/SNP/) (except for dbSNP variants that are registered as clinically associated ([ClinVar](http://www.ncbi.nlm.nih.gov/clinvar/)/[Database of Curated Mutations](http://docm.genome.wustl.edu)), or registered as somatic in [COSMIC v78](http://cancer.sanger.ac.uk/cosmic/))

__Note__ : The filtered set of calls may still harbour germline variants, although to a much lesser extent compared to the unfiltered set. No filtering has been made with respect to sequencing depth.

Here, the filtered, coding variant set is presented. We focus on those that are clinically relevant and those that are most likely to contribute to tumorigenesis. __Tier 1__ constitute variants recorded as prognostic/diagnostic/drug sensitivity biomarkers in the [CIViC database](http://civic.genome.wustl.edu) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers). __Tier 2__ includes coding variants that are found in known mutational hotspots, predicted as cancer driver mutations, or curated as disease-causing. __Tier 3__ includes other coding variants found in oncogenes, tumor suppressor genes, or cancer census genes. __Tier 4__ includes other coding variants (not shown here).

### Knowledge/data sources used for annotation
```{r conditional_print0, child='data_sources.Rmd', eval = show_data_sources}
```

## Results - SNVs/InDels

### Filtering stats

### Filtered call statistics

```{r out.width=c('400px', '400px'), fig.show='hold', echo=F}
print(report_data$sample_stats_plot_coding)
```

### Tier 1 - genomic biomarkers for diagnosis, prognosis, and drug sensitivity {.tabset}

- A total of __`r nrow(report_data$variants_tier1)`__ unique variant(s) in the tumor sample can be mapped to mutation-specific genomic biomarkers in the [database for clinical interpretations of variants in cancer, CIViC](https://civic.genome.wustl.edu/#/home), with the following number of evidence items:
    * Tier 1A: __`r nrow(report_data$clinical_evidence_items_tier1A)`__ evidence items linked to specific amino acid changes
    * Tier 1B: __`r nrow(report_data$clinical_evidence_items_tier1B)`__ evidence items linked to codon alterations
    * Tier 1C: __`r nrow(report_data$clinical_evidence_items_tier1C)`__ evidence items linked to exon alterations

<br>
```{r conditional_print1, child='tier1.Rmd', eval = tier1_report}
```

### Tier 2 - other cancer mutation hotspots, curated mutatations, or predicted driver mutations

- A total of __`r nrow(report_data$variants_tier2)`__ unique variant(s) in the tumor sample are curated as disease-causing, predicted as driver mutations, or occur in previously detected mutation hotspots
<br>
```{r conditional_print2, child='tier2.Rmd', eval = tier2_report}
```

### Tier 3 - other coding mutations in proto-oncogenes/tumor suppressors/cancer census genes

- A total of __`r nrow(report_data$variants_tier3)`__ unique variant(s) in the tumor sample are found within known proto-oncogenes, tumor suppressor genes, or cancer census genes:
<br>
```{r conditional_print3, child='tier3.Rmd', eval = tier3_report}
```


## References
